tools/audio/xml.c

```diff
@@ -0,0 +1,393 @@
+    if (value == NULL || value[0] == '\0')
+        goto err;
+
+    bool neg = false;
+    int res;
+
+    size_t value_len = strlen(value);
+    size_t start;
+
+    // consume initial whitespace
+    for (start = 0; start < value_len; start++) {
+        if (!isspace(value[start]))
+            break;
+    }
+    // if we consumed the whole string, it was bad
+    if (start == value_len)
+        goto err;
+
+    // handle sign character if present
+    if (value[start] == '+' || value[start] == '-') {
+        neg = value[start] == '-';
+        start++;
+    }
+
+    int base;
+
+    // get absolute value in either base 10 or 16
+    if (start + 2 < value_len && value[start + 0] == '0' && tolower(value[start + 1]) == 'x') {
+        start += 2;
+        base = 16;
+    } else {
+        base = 10;
+    }
+
+    char *str_end;
+    res = strtol(&value[start], &str_end, base);
+    size_t end = str_end - value;
+    if (start == end)
+        goto err;
+
+    // consume trailing whitespace
+    while (value[end] != '\0') {
+        if (!isspace(value[end]))
+            goto err;
+        end++;
+    }
+
+    assert(end == value_len);
+
+    // apply sign
+    return neg ? -res : res;
+err:
+    error("bad int value %s", value);
+}
+
+void
+xml_parse_int(const char *value, void *out)
+{
+    int v = xml_str_to_int(value);
+
+    copy_out(out, v);
+}
+
+void
+xml_parse_uint(const char *value, void *out)
+{
+    int v = xml_str_to_int(value);
+    if (v < 0)
+        error("Value should be unsigned");
+
+    copy_out(out, v);
+}
+
+void
+xml_parse_s16(const char *value, void *out)
+{
+    int v = xml_str_to_int(value);
+    if (v < INT16_MIN || v > INT16_MAX)
+        error("Value %d out of range for s16", v);
+    int16_t vs16 = v;
+
+    copy_out(out, vs16);
+}
+
+void
+xml_parse_u8(const char *value, void *out)
+{
+    int v = xml_str_to_int(value);
+    if (v < 0 || v > UINT8_MAX)
+        error("Value %d out of range for u8", v);
+    uint8_t vu8 = v;
+
+    copy_out(out, vu8);
+}
+
+void
+xml_parse_s8(const char *value, void *out)
+{
+    int v = xml_str_to_int(value);
+    if (v < INT8_MIN || v > INT8_MAX)
+        error("Value %d out of range for s8", v);
+    int8_t vs8 = v;
+
+    copy_out(out, vs8);
+}
+
+/**
+ * Parse a note number name to its s8 [0;127] value.
+ * For example "PITCH_EF4" -> 42.
+ */
+void
+xml_parse_note_number(const char *value, void *out)
+{
+    size_t value_len = strlen(value);
+    int8_t vs8;
+    char c;
+
+    if (value_len == 0)
+        goto err;
+
+    // consume initial whitespace
+    size_t start;
+    for (start = 0; start < value_len; start++) {
+        if (!isspace(value[start]))
+            break;
+    }
+    // if we consumed the whole string, it was bad
+    if (start == value_len)
+        goto err;
+
+    c = toupper(value[start]);
+
+    if (c >= 'A' && c <= 'G') {
+        start++;
+
+        // got a note number
+        static const int8_t notes_lut[] = {
+            /* A */ 12, // 9,
```

Comment:
Out of curiosity, what are the commented out numbers for?

Reply:
I had experimented with a couple of approaches for converting note numbers, I don't remember exactly but this is probably left over from one of those experiments. I've removed them since they're not meaningful anymore